### PR TITLE
Use Select: Fix render queue.

### LINKED ISSUE #19199: Different timing of updates between withSelect and useSelect
I'm not sure if this is a bug but I noticed that `withSelect()` and `useSelect()` call events/hooks at different times.

I have a block that shows the number of blocks. When I use `withSelect()` the number of blocks are updated immediately, but if I use `useSelect()` and delete a block with the mouse, then the update happens only after I click somewhere else in the post. See the two examples below:

**My `withSelect()` code**
(updates are immediately coming through)

```
const MyBlock = function(props) {
    return <div>Number of blocks: {props.count}</div>;
};

const MyBlockWithSelect = withSelect((select) => {
    return {
        count: select('core/block-editor').getBlockCount()
    };
})(MyBlock);
```

**My `useSelect()` code**
(delete a block with the mouse triggers no update, only after clicking somewhere else in the post)

```
const MyBlockUseSelect = function(props) {
    const {count} = useSelect((select) => {
        return {
            count: select('core/block-editor').getBlockCount()
        };
    }, []);
    return <div>Number of blocks: {count}</div>;
};
```
Is this intended behavior or am I missing something?

### PR DESCRIPTION
**Pull request by epiqueras**
Depends on #19282
Fixes #19199

This PR fixes the issue described in #19199 and debugged here:

- https://github.com/WordPress/gutenberg/issues/19199#issuecomment-568090443
- https://github.com/WordPress/gutenberg/issues/19199#issuecomment-568105378

### CODE REVIEW

**Review by aduth — APPROVED** [review]
This looks good and makes sense. Nice debugging work 👍 

Two minor (non-blocking) thoughts:

- Is there a test case we could introduce to verify against a regression?
- Is there any _downside_ to using `useMemoOne` vs. the native `useMemo`? I would think there must be some reason that React developers chose not to make the semantic guarantee. Most pressing to me might be if it has any potential performance implications.

**Comment by epiqueras** [comment]
Yes, we can test that going from async to sync mode runs queued updates
correctly.

The native one reserves the right to bust the cache for when React thinks
its wise to free memory. The cache stable one would only be problematic in
cases where a significant amount of memory is not garbage collected because
of it and it’s not even needed, e.g. if the components are off-screen.

**Comment by aduth** [comment]
> Depends on #19282

#19282 was merged today.

I'm trying to recall, was one or both of these going to help unblock #19007, with regards to normalizing how `isAsync` is treated?

**Comment by epiqueras** [comment]
The combination of both.

LGTM!

File: package-lock.json [hunk 1]
```diff
@@ -24718,7 +24719,7 @@
 					"dependencies": {
 						"kind-of": {
 							"version": "2.0.1",
-							"resolved": "https://registry.npmjs.org/kind-of/-/kind-of-2.0.1.tgz",
+							"resolved": "http://registry.npmjs.org/kind-of/-/kind-of-2.0.1.tgz",
```
Comment (aduth) [inline package-lock.json]: When I run `npm install` while on this branch, I have local changes which apply this back to the original:

```diff
diff --git a/package-lock.json b/package-lock.json
index faa74e088c..f7c0903b36 100644
--- a/package-lock.json
+++ b/package-lock.json
@@ -24685,7 +24685,7 @@
 			"dependencies": {
 				"clone-deep": {
 					"version": "0.2.4",
-					"resolved": "http://registry.npmjs.org/clone-deep/-/clone-deep-0.2.4.tgz",
+					"resolved": "https://registry.npmjs.org/clone-deep/-/clone-deep-0.2.4.tgz",
 					"integrity": "sha1-TnPdCen7lxzDhnDF3O2cGJZIHMY=",
 					"dev": true,
 					"requires": {
@@ -24719,7 +24719,7 @@
 					"dependencies": {
 						"kind-of": {
 							"version": "2.0.1",
-							"resolved": "http://registry.npmjs.org/kind-of/-/kind-of-2.0.1.tgz",
+							"resolved": "https://registry.npmjs.org/kind-of/-/kind-of-2.0.1.tgz",
 							"integrity": "sha1-AY7HpM5+OobLkUG+UZ0kyPqpgbU=",
 							"dev": true,
 							"requires": {

```

Related to: https://github.com/WordPress/gutenberg/pull/19680#pullrequestreview-344719610

I'm using latest version of NPM (6.13.6).

Comment (epiqueras) [reply package-lock.json]: Can you push them?

Comment (epiqueras) [reply package-lock.json]: This happens often when contributors merge PRs with different NPM versions.

**Pull request merged**